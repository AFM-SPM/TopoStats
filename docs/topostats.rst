topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
