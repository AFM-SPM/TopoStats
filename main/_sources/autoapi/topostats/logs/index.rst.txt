topostats.logs
==============

.. py:module:: topostats.logs

.. autoapi-nested-parse::

   
   Initialise login module.
















   ..
       !! processed by numpydoc !!


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/topostats/logs/logs/index


