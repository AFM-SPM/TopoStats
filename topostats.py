"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
