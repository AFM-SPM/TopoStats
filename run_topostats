#!/usr/bin/env python3
"""Topotracing"""
import argparse as arg
from asyncio.log import logger
from collections import defaultdict

# from email.policy import default
from functools import partial
from multiprocessing import Pool
from pathlib import Path
from typing import Union, Dict

import pandas as pd
import numpy as np
from tqdm import tqdm
import numpy as np

from topostats.filters import Filters
from topostats.grains import Grains
from topostats.grainstats import GrainStats, get_grainstats
from topostats.io import read_yaml
from topostats.logs.logs import setup_logger, LOGGER_NAME
from topostats.plottingfuncs import plot_and_save
from topostats.thresholds import threshold
from topostats.tracing.dnatracing import dnaTrace
from topostats.utils import convert_path, find_images, update_config

LOGGER = setup_logger(LOGGER_NAME)


def create_parser() -> arg.ArgumentParser:
    """Create a parser for reading options."""
    parser = arg.ArgumentParser(
        description="Process AFM images. Additional arguments over-ride those in the configuration file."
    )
    parser.add_argument(
        "-c",
        "--config_file",
        dest="config_file",
        required=True,
        help="Path to a YAML configuration file.",
    )
    parser.add_argument(
        "-b",
        "--base_dir",
        dest="base_dir",
        type=str,
        required=False,
        help="Base directory to scan for images.",
    )
    parser.add_argument(
        "-f",
        "--file_ext",
        dest="file_ext",
        type=str,
        required=False,
        help="File extension to scan for.",
    )
    parser.add_argument(
        "--channel",
        dest="channel",
        type=str,
        required=False,
        help="Channel to extract.",
    )
    parser.add_argument(
        "-o",
        "--output_dir",
        dest="output_dir",
        type=str,
        required=False,
        help="Output directory to write results to.",
    )
    parser.add_argument(
        "-s",
        "--save_plots",
        dest="save_plots",
        type=bool,
        required=False,
        help="Whether to save plots.",
    )
    parser.add_argument(
        "-a",
        "--amplify_level",
        dest="amplify_level",
        type=float,
        required=False,
        help="Amplify signals by the given factor.",
    )
    parser.add_argument(
        "-t",
        "--threshold_method",
        dest="threshold_method",
        required=False,
        help="Method used for thresholding.",
    )
    parser.add_argument(
        "--otsu_threshold_multiplier",
        dest="otsu_threshold_multiplier",
        required=False,
        help="Factor to scale threshold during grain finding.",
    )
    parser.add_argument("-m", "--mask", dest="mask", type=bool, required=False, help="Mask the image.")
    parser.add_argument(
        "-q",
        "--quiet",
        dest="quiet",
        type=bool,
        required=False,
        help="Toggle verbosity.",
    )
    parser.add_argument(
        "-w",
        "--warnings",
        dest="quiet",
        type=str,
        required=False,
        help="Whether to ignore warnings.",
    )
    return parser


def process_scan(
    image_path: Union[str, Path] = None,
    channel: str = "Height",
    amplify_level: float = 1.0,
    filter_threshold_method: str = "otsu",
    filter_otsu_threshold_multiplier: Union[int, float] = 1.7,
    filter_threshold_std_dev=1.0,
    filter_threshold_abs_lower=None,
    filter_threshold_abs_upper=None,
    gaussian_size: Union[int, float] = 2,
    gaussian_mode: str = "nearest",
    absolute_smallest_grain_size=None,
    background: float = 0.0,
    grains_threshold_method: str = "otsu",
    grains_otsu_threshold_multiplier: Union[int, float] = 1.7,
    grains_threshold_std_dev=1.0,
    grains_threshold_abs_lower=None,
    grains_threshold_abs_upper=None,
    save_plots: bool = True,
    output_dir: Union[str, Path] = "output",
) -> None:
    """Process a single image, filtering, finding grains and calculating their statistics.

    Parameters
    ----------
    image_path : Union[str, Path]
        Path to image to process.
    channel : str
        Channel to extract and process, default 'height'.
    amplify_level : float
        Level to amplify image prior to processing by.
    threshold_method: str
        Method for determining threshold to mask values, default is 'otsu'.
    gaussian_size : Union[int, float]
        Minimum grain size in nanometers (nm).
    mode : str
        Mode for filtering (default is 'nearest').
    otsu_threshold_multiplier : Union[int, float]
        Factor by which lower threshold is to be scaled prior to masking.
    background : float
    save_plots : bool
        Flag as to whether to save plots to PNG files.
    output_dir : Union[str, Path]
        Path to output directory for saving results.

    Examples
    --------

    from topostats.topotracing import process_scan

    process_scan(image_path='minicircle.spm',
                 save_plots=True,
                 output_dir='output/')

    """
    LOGGER.info(f"Processing : {image_path}")
    # Filter Image :
    #
    # The Filters class has a convenience method that runs the instantiated class in full.
    filtered_image = Filters(
        image_path,
        threshold_method=filter_threshold_method,
        threshold_std_dev=filter_threshold_std_dev,
        threshold_absolute_lower=filter_threshold_abs_lower,
        threshold_absolute_upper=filter_threshold_abs_upper,
        channel=channel,
        amplify_level=amplify_level,
        output_dir=output_dir / image_path.stem / "filters",
    )
    filtered_image.filter_image()

    # Find Grains :
    #
    # The Grains class also has a convenience method that runs the instantiated class in full.
    grains = Grains(
        image=filtered_image.images["zero_averaged_background"],
        filename=filtered_image.filename,
        pixel_to_nm_scaling=filtered_image.pixel_to_nm_scaling,
        gaussian_size=gaussian_size,
        gaussian_mode=gaussian_mode,
        threshold_method=grains_threshold_method,
        otsu_threshold_multiplier=grains_otsu_threshold_multiplier,
        threshold_std_dev=grains_threshold_std_dev,
        threshold_absolute_lower=grains_threshold_abs_lower,
        threshold_absolute_upper=grains_threshold_abs_upper,
        absolute_smallest_grain_size=absolute_smallest_grain_size,
        background=background,
        base_output_dir=output_dir / filtered_image.filename / "grains",
    )
    grains.find_grains()

    # Grainstats :
    #
    # Ditto for the GrainStats class.
    grainstats = GrainStats(
        data=grains.images["gaussian_filtered"],
        labelled_data=grains.images["labelled_regions"],
        pixel_to_nanometre_scaling=filtered_image.pixel_to_nm_scaling,
        img_name=filtered_image.filename,
        output_dir=output_dir,
    )
    grain_statistics = grainstats.calculate_stats()

    # Run dnatracing
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_PIXELS.csv"), filtered_image.images["pixels"], delimiter=",")
    # np.savetxt(Path("/home/neil/tmp/TopoStats/TOPOSTATS_GRAINS.csv"), grains.images["labelled_regions"], delimiter=",")
    # np.savetxt(
    #     Path("/home/neil/tmp/TopoStats/TOPOSTATS_GAUSSIAN_FILTERED.csv"),
    #     grains.images["gaussian_filtered"],
    #     delimiter=",",
    # )
    print(f"FROM run_topostats : {grains.images['labelled_regions']}")
    #    print(f"FROM run_topostats : {grains.images['labelled_regions'].tolist()}")
    dna_traces = dnaTrace(
        full_image_data=grains.images["gaussian_filtered"].T,
        grains=grains.images["labelled_regions"],
        afm_image_name=filtered_image.filename,
        pixel_size=filtered_image.pixel_to_nm_scaling,
        number_of_columns=grains.images["labelled_regions"].shape[0],
        number_of_rows=grains.images["labelled_regions"].shape[1],
    )

    # Optionally plot all stages
    if save_plots:
        # Filtering stage
        for plot_name, array in filtered_image.images.items():
            if plot_name not in ["scan_raw", "extracted_channel"]:
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
        # Grain stage - only if we have grains
        if len(grains.region_properties) > 0:
            for plot_name, array in grains.images.items():
                PLOT_DICT[plot_name]["output_dir"] = Path(output_dir) / filtered_image.filename
                plot_and_save(array, **PLOT_DICT[plot_name])
            # Make a plot of coloured regions with bounding boxes
            plot_and_save(
                grains.images["coloured_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["bounding_boxes"],
                region_properties=grains.region_properties,
            )
            plot_and_save(
                grains.images["labelled_regions"],
                Path(output_dir) / filtered_image.filename,
                **PLOT_DICT["coloured_boxes"],
                region_properties=grains.region_properties,
            )
        # Grainstats
        # FIXME : Include this


def main():
    """Run processing."""

    # Parse command line options, load config and update with command line options
    parser = create_parser()
    args = parser.parse_args()
    config = read_yaml(args.config_file)
    config = update_config(config, args)

    LOGGER.info(f"Configuration file loaded from    : {args.config_file}")
    LOGGER.info(f'Scanning for images in            : {config["base_dir"]}')
    LOGGER.info(f'Output directory                  : {config["output_dir"]}')
    LOGGER.info(f'Looking for images with extension : {config["file_ext"]}')
    img_files = find_images(config["base_dir"])
    LOGGER.info(f'Images with extension {config["file_ext"]} in {config["base_dir"]} : {len(img_files)}')

    if config["quiet"]:
        LOGGER.setLevel("ERROR")
    # For debugging (as Pool makes it hard to find things when they go wrong)
    # for x in img_files:
    #     process_scan(
    #         image_path=x,
    #         channel=config["channel"],
    #         amplify_level=config["amplify_level"],
    #         threshold_method=config["threshold_method"],
    #         otsu_threshold_multiplier=config["grains"]["otsu_threshold_multiplier"],
    #         gaussian_size=config["grains"]["gaussian_size"],
    #         gaussian_mode=config["grains"]["gaussian_mode"],
    #         background=config["grains"]["background"],
    #         output_dir=config["output_dir"],
    #     )
    processing_function = partial(
        process_scan,
        channel=config["channel"],
        amplify_level=config["amplify_level"],
        filter_threshold_method=config["filter"]["threshold"]["method"],
        filter_otsu_threshold_multiplier=config["filter"]["threshold"]["otsu_multiplier"],
        filter_threshold_std_dev=config["filter"]["threshold"]["std_dev"],
        filter_threshold_abs_lower=config["filter"]["threshold"]["absolute"][0],
        filter_threshold_abs_upper=config["filter"]["threshold"]["absolute"][1],
        gaussian_size=config["grains"]["gaussian_size"],
        gaussian_mode=config["grains"]["gaussian_mode"],
        absolute_smallest_grain_size=config["grains"]["absolute_smallest_grain_size"],
        background=config["grains"]["background"],
        output_dir=config["output_dir"],
        grains_threshold_method=config["grains"]["threshold"]["method"],
        grains_otsu_threshold_multiplier=config["grains"]["threshold"]["otsu_multiplier"],
        grains_threshold_std_dev=config["grains"]["threshold"]["std_dev"],
        grains_threshold_abs_lower=config["grains"]["threshold"]["absolute"][0],
        grains_threshold_abs_upper=config["grains"]["threshold"]["absolute"][1],
    )

    with Pool(processes=config["cores"]) as pool:
        with tqdm(
            total=len(img_files),
            desc=f'Processing images from {config["base_dir"]}, results are under {config["output_dir"]}',
        ) as pbar:
            for _ in pool.imap_unordered(processing_function, img_files):
                pbar.update()


if __name__ == "__main__":
    main()
