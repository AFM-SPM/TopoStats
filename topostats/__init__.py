"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
