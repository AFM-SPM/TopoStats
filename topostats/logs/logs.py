"""
Standardise logging.
"""
import sys
from pathlib import Path
from datetime import datetime
import logging

# pylint: disable=assignment-from-no-return

start = datetime.now()
LOG_INFO_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] %(message)s", datefmt="%a, %d %b %Y %H:%M:%S"
)
LOG_ERROR_FORMATTER = logging.Formatter(
    fmt="[%(asctime)s] [%(levelname)-8s] [%(name)s] [%(filename)s] [%(lineno)s] %(message)s",
    datefmt="%a, %d %b %Y %H:%M:%S",
)

LOGGER_NAME = "topostats"


def setup_logger(log_name: str = LOGGER_NAME) -> logging.Logger:
    """Setup a standard logger.

    The logger for the module is initialised when the module is loaded (as this functions is called from
    __init__.py). This creates two stream handlers, one for general output and one for errors which are formatted
    differently (there is greater information in the error formatter). To use in modules import the 'LOGGER_NAME' and
    create a logger as shown in the Examples, it will inherit the formatting and direction of messages to the correct
    stream.

    Parameters
    ----------
    log_name : str
        Name under which logging information occurs.

    Returns
    -------
    logging.Logger
        Logger object.

    Examples
    --------

    To use the logger in (sub-)modules have the following.

        import logging
        from topostats.logs.logs import LOGGER_NAME

        LOGGER = logging.getLogger(LOGGER_NAME)

        LOGGER.info('This is a log message.')
    """
    out_stream_handler = logging.StreamHandler(sys.stdout)
    out_stream_handler.setLevel(logging.DEBUG)
    out_stream_handler.setFormatter(LOG_INFO_FORMATTER)

    err_stream_handler = logging.StreamHandler(sys.stderr)
    err_stream_handler.setLevel(logging.ERROR)
    err_stream_handler.setFormatter(LOG_ERROR_FORMATTER)

    file_handler = logging.FileHandler(Path().cwd().stem + f"-{start.strftime('%Y-%m-%d-%H-%M-%S')}.log")
    file_handler.setFormatter(LOG_ERROR_FORMATTER)

    logger = logging.getLogger(log_name)
    logger.setLevel(logging.INFO)
    logger.propagate = True
    if not logger.handlers:
        logger.addHandler(out_stream_handler)
        logger.addHandler(err_stream_handler)
        logger.addHandler(file_handler)

    return logger
