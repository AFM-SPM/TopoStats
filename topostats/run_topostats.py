"""Run TopoStats

This provides an entry point for running TopoStats as a command line programme.
"""
import argparse as arg
from collections import defaultdict
from functools import partial
import importlib.resources as pkg_resources
import logging
from multiprocessing import Pool
from pprint import pformat
import sys
import yaml

import pandas as pd
from tqdm import tqdm

from topostats import __version__
from topostats.io import find_files, read_yaml, save_folder_grainstats, write_yaml, LoadScans
from topostats.logs.logs import LOGGER_NAME
from topostats.plotting import toposum
from topostats.processing import check_run_steps, completion_message, process_scan
from topostats.utils import update_config, update_plotting_config
from topostats.validation import validate_config, DEFAULT_CONFIG_SCHEMA, PLOTTING_SCHEMA, SUMMARY_SCHEMA

# We already setup the logger in __init__.py and it is idempotent so calling it here returns the same object as from
# __init__.py
# Ref : https://stackoverflow.com/a/57799639/1444043
# LOGGER = setup_logger(LOGGER_NAME)
LOGGER = logging.getLogger(LOGGER_NAME)


# pylint: disable=too-many-branches
# pylint: disable=too-many-locals
# pylint: disable=too-many-statements
# pylint: disable=unnecessary-dict-index-lookup
# pylint: disable=too-many-nested-blocks


def create_parser() -> arg.ArgumentParser:
    """Create a parser for reading options."""
    parser = arg.ArgumentParser(
        description="Process AFM images. Additional arguments over-ride those in the configuration file."
    )
    parser.add_argument(
        "-c",
        "--config_file",
        dest="config_file",
        required=False,
        help="Path to a YAML configuration file.",
    )
    parser.add_argument(
        "--create-config-file",
        dest="create_config_file",
        type=str,
        required=False,
        help="Filename to write a sample YAML configuration file to (should end in '.yaml').",
    )
    parser.add_argument(
        "-s",
        "--summary_config",
        dest="summary_config",
        required=False,
        help="Path to a YAML configuration file for summary plots and statistics.",
    )
    parser.add_argument(
        "-b",
        "--base_dir",
        dest="base_dir",
        type=str,
        required=False,
        help="Base directory to scan for images.",
    )
    parser.add_argument(
        "-j",
        "--cores",
        dest="cores",
        type=int,
        required=False,
        help="Number of CPU cores to use when processing.",
    )
    parser.add_argument(
        "-l",
        "--log_level",
        dest="log_level",
        type=str,
        required=False,
        help="Logging level to use, default is 'info' for verbose output use 'debug'.",
    )
    parser.add_argument(
        "-f",
        "--file_ext",
        dest="file_ext",
        type=str,
        required=False,
        help="File extension to scan for.",
    )
    parser.add_argument(
        "--channel",
        dest="channel",
        type=str,
        required=False,
        help="Channel to extract.",
    )
    parser.add_argument(
        "-o",
        "--output_dir",
        dest="output_dir",
        type=str,
        required=False,
        help="Output directory to write results to.",
    )
    parser.add_argument(
        "--save_plots",
        dest="save_plots",
        type=bool,
        required=False,
        help="Whether to save plots.",
    )
    parser.add_argument("-m", "--mask", dest="mask", type=bool, required=False, help="Mask the image.")
    parser.add_argument(
        "-v",
        "--version",
        action="version",
        version=f"Installed version of TopoStats : {__version__}",
        help="Report the current version of TopoStats that is installed.",
    )
    parser.add_argument(
        "-w",
        "--warnings",
        dest="warnings",
        type=bool,
        required=False,
        help="Whether to ignore warnings.",
    )
    return parser


def main(args=None):
    """Find and process all files."""

    # Parse command line options, load config (or default) and update with command line options
    parser = create_parser()
    args = parser.parse_args() if args is None else parser.parse_args(args)
    if args.config_file is not None:
        config = read_yaml(args.config_file)
    else:
        default_config = pkg_resources.open_text(__package__, "default_config.yaml")
        config = yaml.safe_load(default_config.read())
    config = update_config(config, args)

    # Set logging level
    if config["log_level"] == "warning":
        LOGGER.setLevel("WARNING")
    elif config["log_level"] == "error":
        LOGGER.setLevel("ERROR")
    elif config["log_level"] == "debug":
        LOGGER.setLevel("DEBUG")
    else:
        LOGGER.setLevel("INFO")
    # Validate configuration
    validate_config(config, schema=DEFAULT_CONFIG_SCHEMA, config_type="YAML configuration file")

    # Write sample configuration if asked to do so and exit
    if args.create_config_file:
        write_yaml(
            config,
            output_dir="./",
            config_file=args.create_config_file,
            header_message="Sample configuration file auto-generated",
        )
        LOGGER.info(f"A sample configuration has been written to : ./{args.create_config_file}")
        LOGGER.info(
            "Please refer to the documentation on how to use the configuration file : \n\n"
            "https://afm-spm.github.io/TopoStats/usage.html#configuring-topostats\n"
            "https://afm-spm.github.io/TopoStats/configuration.html"
        )
        sys.exit()

    # Create base output directory
    config["output_dir"].mkdir(parents=True, exist_ok=True)

    # Load plotting_dictionary and validate
    plotting_dictionary = pkg_resources.open_text(__package__, "plotting_dictionary.yaml")
    config["plotting"]["plot_dict"] = yaml.safe_load(plotting_dictionary.read())
    validate_config(
        config["plotting"]["plot_dict"], schema=PLOTTING_SCHEMA, config_type="YAML plotting configuration file"
    )

    # Check earlier stages of processing are enabled for later.
    check_run_steps(
        filter_run=config["filter"]["run"],
        grains_run=config["grains"]["run"],
        grainstats_run=config["grainstats"]["run"],
        dnatracing_run=config["dnatracing"]["run"],
    )
    # Update the config["plotting"]["plot_dict"] with plotting options
    config["plotting"] = update_plotting_config(config["plotting"])

    LOGGER.info(f"Configuration file loaded from      : {args.config_file}")
    LOGGER.info(f"Scanning for images in              : {config['base_dir']}")
    LOGGER.info(f"Output directory                    : {str(config['output_dir'])}")
    LOGGER.info(f"Looking for images with extension   : {config['file_ext']}")
    img_files = find_files(config["base_dir"], file_ext=config["file_ext"])
    LOGGER.info(f"Images with extension {config['file_ext']} in {config['base_dir']} : {len(img_files)}")
    if len(img_files) == 0:
        LOGGER.error(f"No images with extension {config['file_ext']} in {config['base_dir']}")
        LOGGER.error("Please check your configuration and directories.")
        sys.exit()
    LOGGER.info(f'Thresholding method (Filtering)     : {config["filter"]["threshold_method"]}')
    LOGGER.info(f'Thresholding method (Grains)        : {config["grains"]["threshold_method"]}')
    LOGGER.debug(f"Configuration after update         : \n{pformat(config, indent=4)}")  # noqa : T203

    processing_function = partial(
        process_scan,
        base_dir=config["base_dir"],
        filter_config=config["filter"],
        grains_config=config["grains"],
        grainstats_config=config["grainstats"],
        dnatracing_config=config["dnatracing"],
        plotting_config=config["plotting"],
        output_dir=config["output_dir"],
    )

    all_scan_data = LoadScans(img_files, **config["loading"])
    all_scan_data.get_data()
    scan_data_dict = all_scan_data.img_dict

    with Pool(processes=config["cores"]) as pool:
        results = defaultdict()
        with tqdm(
            total=len(img_files),
            desc=f"Processing images from {config['base_dir']}, results are under {config['output_dir']}",
        ) as pbar:
            for img, result in pool.imap_unordered(
                processing_function,
                scan_data_dict.values(),
            ):
                results[str(img)] = result
                pbar.update()
    try:
        results = pd.concat(results.values())
    except ValueError as error:
        LOGGER.error("No grains found in any images, consider adjusting your thresholds.")
        LOGGER.error(error)

    # Summary Statistics and Plots
    if config["summary_stats"]["run"]:
        # Load summary plots/statistics configuration and validate, location depends on command line args or value in
        # any config file given, if neither are provided the default topostats/summary_config.yaml is loaded
        if args.summary_config is not None:
            summary_config = read_yaml(args.summary_config)
        elif config["summary_stats"]["config"] is not None:
            summary_config = read_yaml(config["summary_stats"]["config"])
        else:
            summary_yaml = pkg_resources.open_text(__package__, "summary_config.yaml")
            summary_config = yaml.safe_load(summary_yaml.read())
        summary_config = update_config(summary_config, args)
        validate_config(summary_config, SUMMARY_SCHEMA, config_type="YAML summarisation config")
        # We never want to load data from CSV as we are using the data that has just been processed.
        summary_config.pop("csv_file")

        # Load variable to label mapping
        plotting_yaml = pkg_resources.open_text(__package__, "var_to_label.yaml")
        summary_config["var_to_label"] = yaml.safe_load(plotting_yaml.read())
        LOGGER.info("[plotting] Default variable to labels mapping loaded.")

        # If we don't have a dataframe or we do and it is all NaN there is nothing to plot
        if isinstance(results, pd.DataFrame) and not results.isna().values.all():
            if results.shape[0] > 1:
                # If summary_config["output_dir"] does not match or is not a sub-dir of config["output_dir"] it
                # needs creating
                summary_config["output_dir"] = config["output_dir"] / "summary_distributions"
                summary_config["output_dir"].mkdir(parents=True, exist_ok=True)
                LOGGER.info(f"Summary plots and statistics will be saved to : {summary_config['output_dir']}")

                # Plot summaries
                summary_config["df"] = results.reset_index()
                toposum(summary_config)
            else:
                LOGGER.warning(
                    "There are fewer than two grains that have been detected, so"
                    " summary plots cannot be made for this image."
                )
        else:
            LOGGER.warning(
                "There are no results to plot, either...\n\n"
                "* you have disabled grains/grainstats/dnatracing.\n"
                "* no grains have been detected across all scans.\n"
                "* there have been errors.\n\n"
                "If you are not expecting to detect grains please consider disabling"
                "grains/grainstats/dnatracing/plotting/summary_stats. If you are expecting to detect grains"
                " please check log-files for further information."
            )
    else:
        summary_config = None

    # Write statistics to CSV if there is data.
    if isinstance(results, pd.DataFrame) and not results.isna().values.all():
        results.reset_index(inplace=True)
        results.set_index(["image", "threshold", "molecule_number"], inplace=True)
        results.to_csv(config["output_dir"] / "all_statistics.csv", index=True)
        save_folder_grainstats(config["output_dir"], config["base_dir"], results)
        results.reset_index(inplace=True)  # So we can access unique image names
        images_processed = len(results["image"].unique())
    else:
        images_processed = 0
        LOGGER.warning("There are no grainstats or dnatracing statistics to write to CSV.")
    # Write config to file
    config["plotting"].pop("plot_dict")
    write_yaml(config, output_dir=config["output_dir"])
    LOGGER.debug(f"Images processed : {images_processed}")
    completion_message(config, img_files, summary_config, images_processed)


if __name__ == "__main__":
    main()
