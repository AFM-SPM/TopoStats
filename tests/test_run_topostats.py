"""Test end-to-end running of topostats."""
import logging
from pathlib import Path

import pytest

from topostats.logs.logs import LOGGER_NAME
from topostats.run_topostats import main as run_topostats_main

BASE_DIR = Path.cwd()


@pytest.mark.parametrize("option", ("-h", "--help"))
def test_run_topostats_main_help(capsys, option) -> None:
    """Test the -h/--help flag to run_topostats."""
    try:
        run_topostats_main([option])
    except SystemExit:
        pass
    assert "Process AFM images." in capsys.readouterr().out


def test_run_topostats_process_all(caplog) -> None:
    """Test run_topostats completes without error when no arguments are given."""
    caplog.set_level(logging.INFO)
    # Explicitly force loading of topostats/default_config.yaml as I couldn't work out how to invoke process_all()
    # without any arguments as it defaults to 'sys.argv' as this is wrapped within pytest it picks up the arguments
    # pytest was invoked with (see thread on StackOverflow at https://stackoverflow.com/a/55260580/1444043)
    run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}"])
    assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
    assert "Successfully Processed^1    : 1 (100.0%)" in caplog.text


def test_run_topostats_process_debug(caplog) -> None:
    """Test run_topostats with debugging and check DEBUG messages are logged"""
    # Set the logging level of the topostats logger
    with caplog.at_level(logging.DEBUG, logger=LOGGER_NAME):
        run_topostats_main(args=["--config", f"{BASE_DIR / 'topostats' / 'default_config.yaml'}", "-l", "debug"])
        assert "Configuration after update         :" in caplog.text
        assert "File extension : .spm" in caplog.text
        assert "Images processed : 1" in caplog.text
        assert "~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~" in caplog.text
