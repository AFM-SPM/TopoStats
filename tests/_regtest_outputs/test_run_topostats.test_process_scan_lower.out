                   centre_x    centre_y  radius_min  radius_max  radius_mean  radius_median  height_min  height_max  height_median  height_mean     volume       area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold  max_feret  min_feret  Contour Lengths  Circular  End to End Distance  Image Name
Molecule Number                                                                                                                                                                                                                                                                                                                                                                
0                312.461649  135.486299    1.544509    6.672074     4.282007       4.278257   -0.811886   -0.437035      -0.574518    -0.580797 -29.197812  50.272012           106.156919                 5.520612                 13.147774               72.583763      2.381579     lower  13.181642   5.520612        65.483617     False              5.96906  minicircle
